javatests/arcs/core/analysis/ExpressionDataFlowAnalyzerTest.kt

```diff
@@ -0,0 +1,370 @@
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+package arcs.core.analysis
+
+import arcs.core.data.expression.PaxelParser
+import com.google.common.truth.Truth.assertThat
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.junit.runners.JUnit4
+
+@RunWith(JUnit4::class)
+class ExpressionDataFlowAnalyzerTest {
+
+  @Test
+  fun literal_number() {
+    val expr = PaxelParser.parse("5")
+
+    val actual = expr.analyze()
+
+    assertThat(actual).isEqualTo(DependencyNode.LITERAL)
+  }
+
+  @Test
+  fun literal_text() {
+    val expr = PaxelParser.parse("'hello'")
+
+    val actual = expr.analyze()
+
+    assertThat(actual).isEqualTo(DependencyNode.LITERAL)
+  }
+
+  @Test
+  fun literal_boolean() {
+    val expr = PaxelParser.parse("false")
+
+    val actual = expr.analyze()
+
+    assertThat(actual).isEqualTo(DependencyNode.LITERAL)
+  }
+
+  @Test
+  fun literal_null() {
+    val expr = PaxelParser.parse("null")
+
+    val actual = expr.analyze()
+
+    assertThat(actual).isEqualTo(DependencyNode.LITERAL)
+  }
+
+  @Test
+  fun single_variable() {
+    val expr = PaxelParser.parse("x")
+
+    val actual = expr.analyze()
+
+    assertThat(actual).isEqualTo(DependencyNode.PrimitiveValue("x"))
+  }
+
+  @Test
+  fun sum_variable_literal() {
+    val expr = PaxelParser.parse("x + 1")
+
+    val actual = expr.analyze()
+
+    assertThat(actual).isEqualTo(DependencyNode.DerivedFrom(DependencyNode.PrimitiveValue("x")))
+  }
+
+  @Test
+  fun variable_field_access() {
+    val expr = PaxelParser.parse("x.foo")
+
+    val actual = expr.analyze()
+
+    assertThat(actual).isEqualTo(DependencyNode.PrimitiveValue("x", "foo"))
+  }
+
+  @Test
+  fun variable_field_access_nested() {
+    val expr = PaxelParser.parse("x.foo.bar")
+
+    val actual = expr.analyze()
+
+    assertThat(actual).isEqualTo(DependencyNode.PrimitiveValue("x", "foo", "bar"))
+  }
+
+  @Test
+  fun variable_field_access_binop() {
+    val expr = PaxelParser.parse("x.foo.bar + y.foo.bar.baz + z.baz.bar")
+
+    val actual = expr.analyze()
+
+    assertThat(actual).isEqualTo(
+      DependencyNode.DerivedFrom(
+        DependencyNode.PrimitiveValue("x", "foo", "bar"),
+        DependencyNode.PrimitiveValue("y", "foo", "bar", "baz"),
+        DependencyNode.PrimitiveValue("z", "baz", "bar")
+      )
+    )
+  }
+
+  @Test
+  fun literal_binop() {
+    val expr = PaxelParser.parse("1 + 2 + 3 + 4 + 5")
+
+    val actual = expr.analyze()
+
+    assertThat(actual).isEqualTo(DependencyNode.LITERAL)
+  }
+
+  @Test
+  fun variable_field_access_binop_literals() {
+    val expr = PaxelParser.parse("x.foo.bar + y.foo.bar.baz + z.baz.bar + 10 + 20")
+
+    val actual = expr.analyze()
+
+    assertThat(actual).isEqualTo(
+      DependencyNode.DerivedFrom(
```

Comment:
I see that `LITERAL` is not explicitly represented in `DerivedFrom`. Do we want to capture that?

@piotrswigon

Reply:
I was arguing against that. I don't anticipate that being useful right now (for claim generation), but if we change our mind that's an easy change.